### PR TITLE
docs: add adr for v1 to v3 controller migration

### PR DESCRIPTION
**Pull request by salasberryfin**
kind/documentation

**What this PR does / why we need it**:

This is a work-in-progress. 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits into logical changes
- [x] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

**Review by alexander-demicev — DISMISSED** [review]

File: docs/adr/0010-migrate-to-v3-cluster-resource.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+- Status: proposed
+- Date: 2024-06-12
+- Authors: @alexander-demicev @salasberryfin
+- Deciders: @Danil-Grigorev @furkatgofurov7 @yiannistri @mjura
+
+## Context
+
+The current implementation of the import process via Turtles is based on creating a `clusters.provisioning.cattle.io` (also referred to as v1 clusters) resource that represents the Rancher cluster. The controller will watch CAPI cluster resources and trigger the creation of a `clusters.provisioning.cattle.io` when a cluster or the cluster's namespace is labeled for import.
+A new controller was introduced recenty that would follow an equivalent procedure for import but would use `clusters.management.cattle.io` (also referred to as v3 clusters) resources to represent Rancher clusters. For now, Turtles continues to use the original controller and the new implementation is sitting behind a disabled feature gate.
```
Comment (Danil-Grigorev) [inline docs/adr/0010-migrate-to-v3-cluster-resource.md:21]: Can this reference previous ADR introducing the controller and the differences this ADR brings? Also the purpose of this ADR to set the FG to true by default. No need to put “For now … disabled feature gate”

File: docs/adr/0010-migrate-to-v3-cluster-resource.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+After some investigation, we realized that this approach can end up causing a Rancher cluster resource duplication. While testing this proof-of-concept, it is important to note the differences between the Rancher cluster resources created when importing a cluster to Rancher:
+- Both v1 and v3 clusters are created to represent every Rancher clusters.
+- In the case of the current controller, we generate a v1 and let Rancher take care of v3.
+- When migrating to the newer controller, we generate a v3 and let Rancher take care of v1.
+
+#### So, what is the difference between both controllers?
+
+The main difference in terms of logic is that, while watching v1 resources, we can use the name of the clusters to identify whether there is an existing resource that represents the CAPI-provisioned cluster.
+
+This means that the resource itself is easily linked to a CAPI cluster via name.
```
Comment (Danil-Grigorev) [inline docs/adr/0010-migrate-to-v3-cluster-resource.md:37]: This was clearly described in previous ADRs. Please link it.

File: docs/adr/0010-migrate-to-v3-cluster-resource.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+Alternatively, when watching v3 clusters in the new controller, the name of the resource is a semi-random string automatically generated by Kubernetes, which forces us to find a different approach to find the resource of interest. Additionally, v3 clusters are not namespaced but global.
+To simplify the watcher on v3 clusters, we add two labels to the resource to reference the `name` and `namespace` of the CAPI cluster it is linked to.
+
+#### And, what is the difference between v3 clusters for each controller?
+
+Now that we now how we link Rancher cluster -> CAPI cluster, we know that the difference between v3 clusters in both controllers is the existence of the CAPI cluster `name` and `namespace` labels.
+
+This means that, if the new controller we'll be migrating to is able to identify v3 clusters with the required labels, we can simply switch controllers and expect everything to be unchanged and Rancher clusters to maintain the same configuration.
```
Comment (Danil-Grigorev) [inline docs/adr/0010-migrate-to-v3-cluster-resource.md:46]: What problems does this approach have? What if a provisioning cluster does not get management cluster created? How the management cluster created by provisioning differs from the one created by us? It seems quite breaking IMO

Comment (salasberryfin) [reply docs/adr/0010-migrate-to-v3-cluster-resource.md:46]: I would say you cannot import a cluster into Rancher with just the provisioning cluster. If there's a provisioning but not a management cluster resource, looks like the original cluster would not be properly imported. What we're saying is that the differences between the mgmt cluster created by provisioning and created by controller (if there are any) are not relevant to the import mechanism, as long as the controller is able to watch and identify the resources (which require the labels and, additionally, a proper display name).

**Comment by salasberryfin** [comment]
Thanks @alexander-demicev @Danil-Grigorev. I added references to the original v3 controller ADR where relevant.

File: docs/adr/0010-migrate-to-v3-cluster-resource.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+#### How does this affect import controllers?
+
+With the current import mechanism, Rancher automatically creates a v3 clusters that links to the v1 cluster generated by Turtles. As stated before, this means that no new resources need to be created for the new import controller to operate normally.
+During migration Turtles needs to be aware of the existing imported clusters and apply labels accordingly so these resources can be reused. As part of the migration guide, we will provide a migration script that allows users to prepare imported clusters for migration in batches, which will simplify the migration process (you can refer to [this PR](https://github.com/rancher/turtles/pull/574)). During this process, there needs to be a mechanism to identify which clusters are actually managed by Turtles: this can simply be done by looking for the `-capi` suffix in the resource name, which is added to all Rancher cluster resources created from Turtles. Additionally, after resources have been prepared for migration, they must be marked to avoid repeating the process on the same resource more than once. An annotation on the `clusters.provisioning.cattle.io` will be used for this:
```
Comment (Danil-Grigorev) [inline docs/adr/0010-migrate-to-v3-cluster-resource.md:53]: Should this also mention that we are matching the clusters using `cluster-api.cattle.io/owned` label under the hood in our controller? It feels like undocumented detail in the ADRs flow, and it takes part in applying `migrated` annotation.

File: docs/adr/0010-migrate-to-v3-cluster-resource.md (hunk 3 above)
Comment (Danil-Grigorev) [reply docs/adr/0010-migrate-to-v3-cluster-resource.md:48]: It is likely that if the rancher version is below 2.8, the management cluster for provisioning might not be created by default. Turtles can be installed there, still, using regular installation path described in the [docs](https://turtles.docs.rancher.com/getting-started/install-rancher-turtles/using_rancher_dashboard).

**Comment by salasberryfin** [comment]
Added reference to automatic migration ADR and `cluster-api.cattle.io/owned` label. I think it is ready for a final review.

**Review by Danil-Grigorev — APPROVED** [review]

**Review by furkatgofurov7 — APPROVED** [review]

**Review by Danil-Grigorev — APPROVED** [review]

**Comment by furkatgofurov7** [comment]
Merging this as this is just a documentation PR and there is no point of running/waiting for E2E tests

**Pull request merged**